packages/core/typescript/TypeScriptFormattingService.ts

```diff
@@ -0,0 +1,217 @@
+import * as ts from 'typescript';
+import { IFormattingService, IFormatSettings } from '../types';
+import { IFileSystem } from '../types/FileSystem';
+import { Util } from '../util';
+
+const NEW_LINE_PLACEHOLDER = '//I keep the new line';
+export class FormattingService implements IFormattingService {
+  private _printer: ts.Printer | undefined;
+  private _formatSettingsFromConfig: IFormatSettings = {};
+  private _defaultFormatSettings: IFormatSettings = {
+    indentSize: 3,
+    tabSize: 4,
+    newLineCharacter: ts.sys.newLine,
+    convertTabsToSpaces: true,
+    indentStyle: ts.IndentStyle.Smart,
+    insertSpaceAfterCommaDelimiter: true,
+    insertSpaceAfterSemicolonInForStatements: true,
+    insertSpaceBeforeAndAfterBinaryOperators: true,
+    insertSpaceAfterKeywordsInControlFlowStatements: true,
+    insertSpaceAfterTypeAssertion: true,
+    singleQuotes: true,
+  };
+
+  /**
+   * The printer instance to use to get the source code from the AST.
+   */
+  private get printer(): ts.Printer {
+    if (!this._printer) {
+      this._printer = ts.createPrinter(this.printerOptions);
+    }
+
+    return this._printer;
+  }
+
+  /**
+   * Create a new formatting service for the given source file.
+   * @param sourceFile The source file to format.
+   * @param fileSystem The file system to use when reading formatting options.
+   * @param formatSettings Custom formatting settings to apply.
+   * @param printerOptions Options to use when printing the source file.
+   * @param compilerOptions Compiler options to use when transforming the source file.
+   */
+  constructor(
+    public sourceFile: ts.SourceFile,
+    private readonly fileSystem?: IFileSystem,
+    private readonly formatSettings?: IFormatSettings,
+    private readonly printerOptions?: ts.PrinterOptions,
+    private readonly compilerOptions?: ts.CompilerOptions
+  ) {}
+
+  /**
+   * Apply formatting to the source file.
+   */
+  public applyFormatting(): string {
+    this.readFormatConfigs();
+    const changes = this.languageService.getFormattingEditsForDocument(
+      this.sourceFile.fileName,
+      this.formatOptions
+    );
+
+    if (this.formatOptions.singleQuotes) {
+      this.sourceFile = ts.transform(
+        this.sourceFile,
+        [this.convertQuotesTransformer],
+        this.compilerOptions
+      ).transformed[0] as ts.SourceFile;
+    }
+
+    const text = this.applyChanges(
+      this.printer.printFile(this.sourceFile),
```

Comment:
Huh? There really is no point re-printing the file over and over just so this service can work with `ts.SourceFile`.
How about we give it the path it needs anyway and the formatting just reads and manipulates the text as intended?

Reply:
I agree that with the current implementation it makes more sense to pass in the `path/to/file` that will use the `App.container` to read and apply formatting.